adsmp/app.py

```diff
@@ -483,3 +486,128 @@ def metrics_delete_by_bibcode(self, bibcode):
                 session.delete(r)
                 session.commit()
                 return True
+            
+    def checksum(self, data):
+        if isinstance(data, basestring):
+            return hex(zlib.crc32(unicode(data)) & 0xffffffff)
+        else:
+            data = deepcopy(data)
+            # remove all the modification timestamps
+            for k,v in data.items():
+                if 'mtime' in k or 'ctime' in k or 'update_timestamp' in k:
+                    del data[k]
+            print hex(zlib.crc32(json.dumps(data, sort_keys=True)) & 0xffffffff), data
+            return hex(zlib.crc32(json.dumps(data, sort_keys=True)) & 0xffffffff)
+    
+    
+    def update_remote_targets(self, solr=None, metrics=None, links=None,
+                              commit_solr=False):
+        """Updates remote databases/solr
+            @param batch: list solr documents (already formatted)
+            @param metrics: tuple with two lists, the first is a list
+                of metric dicts to be inserted; the second is a list
+                of metric dicts to be updated
+            @param links_data: list of dicts to send to the remote
+                resolver links service
+            @return: 
+                (set of processed bibcodes, set of bibcodes that failed)
+                
+            @warning: this call has a side-effect, we are emptying the passed
+                in lists of data (to free up memory)
+        """
+        
+        
+        if metrics and not isinstance(metrics, tuple):
+            raise Exception('Wrong data type passed in for metrics')
+        
+        batch = solr or []
+        batch_insert, batch_update = metrics[0], metrics[1]
+        links_data = links or []
+        
+
+        links_url = self.conf.get('LINKS_RESOLVER_UPDATE_URL')
+        recs_to_process = set()
+        failed_bibcodes = []
+        crcs = {}
+        
+        # take note of all the recs that should be updated
+        # NOTE: we are assuming that set(solr) == set(metrics)_== set(links)
+        for x in (batch, batch_insert, batch_update, links_data):
+            if x:
+                for y in x:
+                    if 'bibcode' in y:
+                        recs_to_process.add(y['bibcode'])
+
+        def update_crc(type, data, faileds):
+            while len(data): # we are freeing memory as well
+                x = data.pop()
+                b = x['bibcode']
+                if b in faileds:
+                    continue
+                crcs.setdefault(b, {})
```

Comment:
thanks, Sergi, definitely helped to make the code better! this is the only place where i'd disagree - we can't set the keys `crcs[b]` because that would override previously computed checksums. as for the locally scoped crcs, yes, i'm being little lazy - but i think that's fine here, it will never go out of the method scope